File: pocs/integrations/src/graph/coherence.ts
import { quotedIdentifier as qi } from "@duckdb/node-api";
import type { QueryableStore } from "../staging/types.js";
import type { GraphClient } from "./types.js";
import type { GraphSinkConfig, LinkPipeline, Step, TablePipeline } from "../transform/pipeline.js";
import { deterministicUuid, compositeEntityId } from "./client.js";
import { readMeta, writeMeta, type MetaScope, type SinkMeta } from "./state-meta.js";
import { escLiteral } from "./sink.js";
import type { Logger } from "../log.js";

export type SinkRef = { sinkId: string; config: GraphSinkConfig };

export function collectGraphSinks(pipelines: readonly TablePipeline[]): SinkRef[] {
  const sinks: SinkRef[] = [];
  const walk = (steps: readonly Step[]) => {
    for (const step of steps) {
      if (step.kind === "graph-sink") sinks.push({ sinkId: step.id, config: step.config });
      if (step.kind === "branch") for (const branch of step.branches) walk(branch as readonly Step[]);
    }
  };
  for (const tp of pipelines) walk(tp.pipeline.steps);
  return sinks;
}

type Fingerprint = { graphIdentity: string; webId: string; namespace: string };

function fingerprintOf(meta: SinkMeta): Fingerprint | null {
  if (meta.graphIdentity == null || meta.webId == null || meta.namespace == null) return null;
  return { graphIdentity: meta.graphIdentity, webId: meta.webId, namespace: meta.namespace };
}

function mismatch(stored: Fingerprint | null, current: Fingerprint): string | null {
  if (!stored) return "state exists but no fingerprint is recorded (state predates the coherence check, or meta was lost)";
  const diffs: string[] = [];
  if (stored.graphIdentity !== current.graphIdentity) diffs.push(`graph "${stored.graphIdentity}" vs "${current.graphIdentity}"`);
  if (stored.webId !== current.webId) diffs.push(`web "${stored.webId}" vs "${current.webId}"`);
  if (stored.namespace !== current.namespace) diffs.push(`namespace "${stored.namespace}" vs "${current.namespace}"`);
  return diffs.length > 0 ? `stored vs current: ${diffs.join(", ")}` : null;
}

const REMEDY =
  "A mismatched target mis-diffs (wrong ops or duplicate entities). Point the state dir at the graph it was " +
  "written against, use a fresh state dir for a clean full sync, or set HASH_ALLOW_STATE_MISMATCH=1 to drop " +
  "local state and cold-start (inserts only, no archives).";

/** Deterministic given the state dir and target graph; retrying cannot succeed. */
export class StateCoherenceError extends Error {
  readonly nonRetryable = true;
}

function allowMismatch(): boolean {
  return process.env.HASH_ALLOW_STATE_MISMATCH === "1";
}

async function stateTablesWithPrefix(db: QueryableStore, prefix: string): Promise<string[]> {
  const { rows } = await db.query(
    `SELECT table_name FROM information_schema.tables WHERE starts_with(table_name, ${escLiteral(prefix)})`,
  );
  return rows.map((r) => String(r.table_name));
}

async function upsertFingerprint(db: QueryableStore, scope: MetaScope, fp: Fingerprint): Promise<void> {
  const existing = await readMeta(db, scope);
  await writeMeta(db, scope, {
    hashVersion: existing?.hashVersion ?? null,
    configHash: existing?.configHash ?? null,
    graphIdentity: fp.graphIdentity,
    webId: fp.webId,
    namespace: fp.namespace,
  });
}

/**
 * Verifies that the local sync state was written against the graph this run targets,
 * before any state or graph write. State diffing is only meaningful against the graph
 * it was built from; a reseeded graph or a different web/namespace re-keys entity ids.
 */
export async function checkStateCoherence(opts: {
  db: QueryableStore;
  client: GraphClient;
  connectorId: string;
  sinks: SinkRef[];
  linkPipelines: readonly LinkPipeline[];
  log?: Logger;
}): Promise<void> {
  const { db, client, connectorId, sinks, linkPipelines, log } = opts;
  const graphIdentity = client.identity();

  const syncTables = new Set(await stateTablesWithPrefix(db, `_state/sync/${connectorId}/`));
  const linkTables = new Set(await stateTablesWithPrefix(db, `_state/links/${connectorId}/`));

  const scoped: { scope: MetaScope; current: Fingerprint; stateTable: string | null; entityType?: string }[] = [
    ...sinks.map((s) => ({
      scope: { scope: "entity" as const, connectorId, sinkId: s.sinkId },
      current: { graphIdentity, webId: s.config.webId, namespace: s.config.idNamespace ?? connectorId },
      stateTable: syncTables.has(`_state/sync/${connectorId}/${s.sinkId}`) ? `_state/sync/${connectorId}/${s.sinkId}` : null,
      entityType: s.config.entityType,
    })),
    ...linkPipelines.map((lp) => ({
      scope: { scope: "link" as const, connectorId, sinkId: lp.id },
      current: { graphIdentity, webId: lp.webId, namespace: lp.idNamespace ?? connectorId },
      stateTable: linkTables.has(`_state/links/${connectorId}/${lp.id}`) ? `_state/links/${connectorId}/${lp.id}` : null,
    })),
  ];

  const failures: string[] = [];
  for (const s of scoped) {
    if (!s.stateTable) continue; // cold start for this sink; fingerprint written below
    const meta = await readMeta(db, s.scope);
    const reason = mismatch(meta ? fingerprintOf(meta) : null, s.current);
    if (reason) failures.push(`${s.scope.scope} sink "${s.scope.sinkId}": ${reason}`);
  }

  if (failures.length === 0) {
    const probeFailure = await sentinelProbe(db, client, scoped, log);
    if (probeFailure) failures.push(probeFailure);
  }

  if (failures.length > 0) {
    if (!allowMismatch()) {
      throw new StateCoherenceError(`state/graph coherence check failed:\n  ${failures.join("\n  ")}\n${REMEDY}`);
    }
    log?.warn(`state/graph coherence OVERRIDDEN (HASH_ALLOW_STATE_MISMATCH=1): dropping state for connector "${connectorId}"`);
    for (const prefix of [`_state/sync/${connectorId}/`, `_state/links/${connectorId}/`, `_state/links-next/${connectorId}/`]) {
      for (const table of await stateTablesWithPrefix(db, prefix)) {
        await db.exec(`DROP TABLE IF EXISTS ${qi(table)}`);
      }
    }
    await db.exec(`DROP TABLE IF EXISTS ${qi(`_state/pending-links/${connectorId}`)}`);
  }

  for (const s of scoped) {
    await upsertFingerprint(db, s.scope, s.current);
  }
}

// One legitimately purged entity must not trip the check; a wiped graph trips it always.
const PROBE_SAMPLE = 3;

async function sentinelProbe(
  db: QueryableStore,
  client: GraphClient,
  scoped: { scope: MetaScope; current: Fingerprint; stateTable: string | null; entityType?: string }[],
  log?: Logger,
): Promise<string | null> {
  const candidates = scoped.filter((s) => s.stateTable && s.entityType);
  if (candidates.length === 0) return null;

  let best: (typeof candidates)[number] | null = null;
  let bestCount = 0;
  for (const c of candidates) {
    const { rows } = await db.query(`SELECT COUNT(*)::BIGINT AS n FROM ${qi(c.stateTable!)}`);
    const n = Number(rows[0]?.n ?? 0);
    if (n > bestCount) {
      best = c;
      bestCount = n;
    }
  }
  if (!best || bestCount === 0) return null;

  const { rows } = await db.query(
    `SELECT _entity_id FROM ${qi(best.stateTable!)} ORDER BY _entity_id LIMIT ${PROBE_SAMPLE}`,
  );
  const ids = rows.map((r) => String(r._entity_id));
  let found = 0;
  for (const id of ids) {
    const composite = compositeEntityId(best.current.webId, deterministicUuid(best.current.namespace, best.entityType!, id));
    if (await client.hasEntity(composite)) found += 1;
  }
  if (found > 0) {
    log?.debug(`coherence probe: ${found}/${ids.length} sentinel entities present`);
    return null;
  }
  return (
    `sentinel probe: none of ${ids.length} sampled entities from "${best.stateTable}" exist in the target graph ` +
    `(graph wiped or reseeded since this state was written?)`
  );
}

File: pocs/integrations_rs/src/graph/coherence.rs
//! State diffing is only meaningful against the graph it was built from.
//! Before any write: per-sink fingerprint {graph identity, web id, namespace}
//! vs `_state/meta`, plus a sentinel probe (three sampled state ids checked
//! in the live graph; all absent means wiped or reseeded). Mismatch is a
//! hard, non-retryable error; HASH_ALLOW_STATE_MISMATCH=1 drops connector
//! state and cold-starts.

use error_stack::{Report, ResultExt as _};
use serde_json::Value;

use crate::build::Integration;
use crate::config::{self, Env};
use crate::error::{CoherenceError, GraphError};
use crate::store::{lit, qi, Store};

use super::state_meta::{self, Meta};
use super::uuid::{composite_entity_id, deterministic_uuid};
use super::SharedClient;

const PROBE_SAMPLE: usize = 3;

/// The two ways a coherence check can fail, kept distinct because they demand
/// opposite handling: a `Mismatch` is deterministic (wrong state dir, or the
/// graph really was wiped) and must NOT be retried; an `Unreachable` means the
/// graph could not be reached during the sentinel probe and MUST be retried
/// with state intact, never mistaken for "graph wiped". Folding a transport
/// error into `Mismatch` would steer the operator toward the state-dropping
/// remedy over a transient blip.
#[derive(Debug)]
pub enum CheckError {
    Mismatch(Report<CoherenceError>),
    Unreachable(Report<GraphError>),
}

impl From<Report<CoherenceError>> for CheckError {
    fn from(report: Report<CoherenceError>) -> Self {
        Self::Mismatch(report)
    }
}

const REMEDY: &str = "A mismatched target mis-diffs (wrong ops or duplicate entities). Point the state dir at the graph it was written against, use a fresh state dir for a clean full sync, or set HASH_ALLOW_STATE_MISMATCH=1 to drop local state and cold-start (inserts only, no archives).";

struct Scoped {
    scope: &'static str,
    sink_id: String,
    graph_identity: String,
    web_id: String,
    namespace: String,
    state_table: Option<String>,
    entity_type: Option<String>,
}

pub async fn check(
    store: &Store,
    client: &SharedClient,
    integration: &Integration,
    env: &Env,
) -> Result<(), CheckError> {
    let connector_id = &integration.connector_id;
    let graph_identity = client.identity();

    let sync_tables = tables_with_prefix(store, &format!("_state/sync/{connector_id}/"))
        .await
        .change_context(CoherenceError)?;
    let link_tables = tables_with_prefix(store, &format!("_state/links/{connector_id}/"))
        .await
        .change_context(CoherenceError)?;

    let mut scoped: Vec<Scoped> = vec![];
    for sink in crate::engine::batch_sync::collect_sinks(&integration.pipelines) {
        let state = format!("_state/sync/{connector_id}/{}", sink.sink_id);
        scoped.push(Scoped {
            scope: "entity",
            sink_id: sink.sink_id.clone(),
            graph_identity: graph_identity.clone(),
            web_id: sink.config.web_id.clone(),
            namespace: sink
                .config
                .id_namespace
                .clone()
                .unwrap_or_else(|| connector_id.clone()),
            state_table: sync_tables.contains(&state).then_some(state),
            entity_type: Some(sink.config.entity_type.clone()),
        });
    }
    for link in &integration.link_pipelines {
        let state = format!("_state/links/{connector_id}/{}", link.id);
        scoped.push(Scoped {
            scope: "link",
            sink_id: link.id.clone(),
            graph_identity: graph_identity.clone(),
            web_id: link.web_id.clone(),
            namespace: link
                .id_namespace
                .clone()
                .unwrap_or_else(|| connector_id.clone()),
            state_table: link_tables.contains(&state).then_some(state),
            entity_type: None,
        });
    }

    let mut failures = vec![];
    for entry in scoped.iter().filter(|entry| entry.state_table.is_some()) {
        let meta = state_meta::read(store, entry.scope, connector_id, &entry.sink_id)
            .await
            .change_context(CoherenceError)?;
        if let Some(reason) = mismatch_reason(meta.as_ref(), entry) {
            failures.push(format!(
                "{} sink \"{}\": {reason}",
                entry.scope, entry.sink_id
            ));
        }
    }

    if failures.is_empty() {
        if let Some(probe_failure) = sentinel_probe(store, client, &scoped).await? {
            failures.push(probe_failure);
        }
    }

    if !failures.is_empty() {
        if config::allow_state_mismatch(env) {
            tracing::warn!(
                "state/graph coherence OVERRIDDEN (HASH_ALLOW_STATE_MISMATCH=1): dropping state for connector \"{connector_id}\""
            );
            for prefix in [
                format!("_state/sync/{connector_id}/"),
                format!("_state/links/{connector_id}/"),
                format!("_state/links-next/{connector_id}/"),
            ] {
                for table in tables_with_prefix(store, &prefix)
                    .await
                    .change_context(CoherenceError)?
                {
                    store
                        .exec(&format!("DROP TABLE IF EXISTS {}", qi(&table)))
                        .await
                        .change_context(CoherenceError)?;
                }
            }
            store
                .exec(&format!(
                    "DROP TABLE IF EXISTS {}",
                    qi(&format!("_state/pending-links/{connector_id}"))
                ))
                .await
                .change_context(CoherenceError)?;
        } else {
            return Err(CheckError::Mismatch(
                Report::new(CoherenceError).attach_printable(format!(
                    "state/graph coherence check failed:\n  {}\n{REMEDY}",
                    failures.join("\n  ")
                )),
            ));
        }
    }

    for entry in &scoped {
        let existing = state_meta::read(store, entry.scope, connector_id, &entry.sink_id)
            .await
            .change_context(CoherenceError)?
            .unwrap_or_default();

        state_meta::write(
            store,
            entry.scope,
            connector_id,
            &entry.sink_id,
            &Meta {
                hash_version: existing.hash_version,
                config_hash: existing.config_hash,
                graph_identity: Some(entry.graph_identity.clone()),
                web_id: Some(entry.web_id.clone()),
                namespace: Some(entry.namespace.clone()),
            },
        )
        .await
        .change_context(CoherenceError)?;
    }

    Ok(())
}

fn mismatch_reason(meta: Option<&Meta>, entry: &Scoped) -> Option<String> {
    const NO_FINGERPRINT: &str =
        "state exists but no fingerprint is recorded (state predates the coherence check, or meta was lost)";

    let Some(meta) = meta else {
        return Some(NO_FINGERPRINT.to_owned());
    };
    let (Some(graph), Some(web), Some(namespace)) =
        (&meta.graph_identity, &meta.web_id, &meta.namespace)
    else {
        return Some(NO_FINGERPRINT.to_owned());
    };

    let diffs: Vec<String> = [
        (graph, &entry.graph_identity, "graph"),
        (web, &entry.web_id, "web"),
        (namespace, &entry.namespace, "namespace"),
    ]
    .into_iter()
    .filter(|(stored, now, _)| stored != now)
    .map(|(stored, now, label)| format!("{label} \"{stored}\" vs \"{now}\""))
    .collect();

    (!diffs.is_empty()).then(|| format!("stored vs current: {}", diffs.join(", ")))
}

// One legitimately purged entity must not trip the check; a wiped graph
// trips it always.
async fn sentinel_probe(
    store: &Store,
    client: &SharedClient,
    scoped: &[Scoped],
) -> Result<Option<String>, CheckError> {
    let mut best: Option<(&Scoped, i64)> = None;
    for entry in scoped {
        let (Some(state_table), Some(_)) = (&entry.state_table, &entry.entity_type) else {
            continue;
        };
        let count = store
            .query(&format!(
                "SELECT COUNT(*)::BIGINT AS n FROM {}",
                qi(state_table)
            ))
            .await
            .change_context(CoherenceError)?
            .single_i64();
        if best.map(|(_, n)| count > n).unwrap_or(true) {
            best = Some((entry, count));
        }
    }

    let Some((entry, count)) = best else {
        return Ok(None);
    };
    if count == 0 {
        return Ok(None);
    }
    let state_table = entry.state_table.as_ref().expect("filtered");
    let entity_type = entry.entity_type.as_ref().expect("filtered");

    let sample = store
        .query(&format!(
            "SELECT _entity_id FROM {} ORDER BY _entity_id LIMIT {PROBE_SAMPLE}",
            qi(state_table)
        ))
        .await
        .change_context(CoherenceError)?;

    let ids: Vec<&str> = sample
        .rows
        .iter()
        .filter_map(|row| row.first().and_then(Value::as_str))
        .collect();

    let mut found = 0;
    for id in &ids {
        let composite = composite_entity_id(
            &entry.web_id,
            &deterministic_uuid(
                &entry.namespace,
                entity_type,
                &Value::String((*id).to_owned()),
            ),
        );
        // A transport error here is NOT "entity absent": propagate it as
        // Unreachable so the run retries with state intact, rather than
        // reading a graph outage as "wiped" and steering toward the
        // state-dropping remedy.
        if client
            .has_entity(&composite)
            .await
            .map_err(CheckError::Unreachable)?
        {
            found += 1;
        }
    }

    if found == 0 {
        Ok(Some(format!(
            "sentinel probe: none of {} sampled entities from \"{state_table}\" exist in the target graph (graph wiped or reseeded since this state was written?)",
            ids.len()
        )))
    } else {
        Ok(None)
    }
}

pub async fn tables_with_prefix(
    store: &Store,
    prefix: &str,
) -> Result<Vec<String>, Report<crate::error::StoreError>> {
    let result = store
        .query(&format!(
            "SELECT table_name FROM information_schema.tables WHERE starts_with(table_name, {})",
            lit(prefix)
        ))
        .await?;

    Ok(result
        .rows
        .iter()
        .filter_map(|row| row.first().and_then(Value::as_str))
        .map(str::to_owned)
        .collect())
}
